src/test/osd/RadosModel.h

```diff
@@ -1398,24 +1403,45 @@ class ReadOp : public TestOp {
       attrretval(0)
   {}
 
+  static std::pair<uint64_t, uint64_t> maybe_randomize_offlen(
+    uint8_t offlen_randomization_ratio,
+    uint64_t max_len)
+  {
+    if ((rand() % 100) < offlen_randomization_ratio && max_len > 0) {
+      // the random offset here is de dacto "first n bytes to skip in
+      // a chhunk" -- it doesn't care about good distrubution across
+      // entire object. imperfect but should be good enough for parital
+      // read testing.
+      const auto off = rand() % max_len;
+      return {off, max_len - off};
+    } else {
+      return {0, max_len};
```

Comment:
We might want to randomize the length as well if `max_len` is always too widely aligned.

Reply:
I think the requirements in this aspect are rather loose:

```cpp
      uint64_t alignment = 0;
      if (requires_alignment) {
        r = context->io_ctx.pool_required_alignment2(&alignment);
        ceph_assert(r == 0);
        ceph_assert(alignment != 0);
      }
      cont_gen = new AppendGenerator(
        prev_length,
        alignment,
        context->min_stride_size,
        context->max_stride_size,
        3);
    } else {
```

In testing both offsets and lengths seem to be quite random (e.g. `read 30517~18635`):
```
[rzarzynski@o06 build]$ CEPH_ARGS="--debug-ms=5 --log-to-stderr=true" bin/ceph_test_rados --no-omap --ec-pool --max-ops 400000 --objects 1024 --max-in-flight 64 --size 400000 --min-stride-size 2 --max-stride-size 8192 --max-seconds 600 --op read 100 --op write 0 --op delete 50 --op snap_create 50 --op snap_remove 50 --op rollback 50 --op setattr 25 --op rmattr 25 --op copy_from 50 --op append 50 --op write_excl 0 --op append_excl 50 --pool unique_pool_1 --offlen_randomization_ratio 100
...
1138: read oid 264 snap -1                                                                                                                                                                                         
1138:  expect (ObjNum 263 snap 0 seq_num 263)                                                                                                                                                                      
2024-03-29T18:24:00.038+0000 7f4218a0ebc0  1 -- 127.0.0.1:0/440189136 --> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] -- osd_op(unknown.0.0:4340 1.as0 1:5feaa37b:::o061534161-264:head [read 36948
~12204,checksum in=4b,getxattrs] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e1844) v9 -- 0x1d63450 con 0x1bf4630                                                                                 
2024-03-29T18:24:00.038+0000 7f4218a0ebc0  5 --2- 127.0.0.1:0/440189136 >> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] conn(0x1bf4630 0x1bf6a50 crc :-1 s=READY pgs=21 cs=0 l=1 rev1=1 crypto rx=0 
tx=0 comp rx=0 tx=0).send_message enqueueing message m=0x1d63450 type=42 osd_op(client.4237.0:4340 1.as0 1:5feaa37b:::o061534161-264:head [read 36948~12204,checksum in=4b,getxattrs] snapc 0=[] ondisk+read+known_
if_redirected+supports_pool_eio e1844) v9                                                                                                                                                                          
2024-03-29T18:24:00.038+0000 7f4218a0ebc0  1 -- 127.0.0.1:0/440189136 --> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] -- osd_op(unknown.0.0:4341 1.as0 1:5feaa37b:::o061534161-264:head [read 3051$
~18635,checksum in=4b] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e1844) v9 -- 0x1cd2400 con 0x1bf4630                                                                                           
2024-03-29T18:24:00.038+0000 7f4212d03700  5 --2- 127.0.0.1:0/440189136 >> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] conn(0x1bf4630 0x1bf6a50 crc :-1 s=READY pgs=21 cs=0 l=1 rev1=1 crypto rx=0 
tx=0 comp rx=0 tx=0).write_message sending message m=0x1d63450 seq=2938 osd_op(client.4237.0:4340 1.as0 1:5feaa37b:::o061534161-264:head [read 36948~12204,checksum in=4b,getxattrs] snapc 0=[] ondisk+read+known_$
f_redirected+supports_pool_eio e1844) v9                                                                                                                                                                           
2024-03-29T18:24:00.038+0000 7f4218a0ebc0  5 --2- 127.0.0.1:0/440189136 >> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] conn(0x1bf4630 0x1bf6a50 crc :-1 s=READY pgs=21 cs=0 l=1 rev1=1 crypto rx=0 
tx=0 comp rx=0 tx=0).send_message enqueueing message m=0x1cd2400 type=42 osd_op(client.4237.0:4341 1.as0 1:5feaa37b:::o061534161-264:head [read 30517~18635,checksum in=4b] snapc 0=[] ondisk+read+known_if_redire$
ted+supports_pool_eio e1844) v9                                                                                                                                                                                    
2024-03-29T18:24:00.038+0000 7f4218a0ebc0  1 -- 127.0.0.1:0/440189136 --> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] -- osd_op(unknown.0.0:4342 1.as0 1:5feaa37b:::o061534161-264:head [read 3195$
~17196,checksum in=4b] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e1844) v9 -- 0x1e6b770 con 0x1bf4630                                                                                           
2024-03-29T18:24:00.038+0000 7f4218a0ebc0  5 --2- 127.0.0.1:0/440189136 >> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] conn(0x1bf4630 0x1bf6a50 crc :-1 s=READY pgs=21 cs=0 l=1 rev1=1 crypto rx=0 
tx=0 comp rx=0 tx=0).send_message enqueueing message m=0x1e6b770 type=42 osd_op(client.4237.0:4342 1.as0 1:5feaa37b:::o061534161-264:head [read 31956~17196,checksum in=4b] snapc 0=[] ondisk+read+known_if_redire$
ted+supports_pool_eio e1844) v9                                                                                                                                                                                    
2024-03-29T18:24:00.038+0000 7f4212d03700  5 --2- 127.0.0.1:0/440189136 >> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] conn(0x1bf4630 0x1bf6a50 crc :-1 s=READY pgs=21 cs=0 l=1 rev1=1 crypto rx=0 
tx=0 comp rx=0 tx=0).write_message sending message m=0x1cd2400 seq=2939 osd_op(client.4237.0:4341 1.as0 1:5feaa37b:::o061534161-264:head [read 30517~18635,checksum in=4b] snapc 0=[] ondisk+read+known_if_redirec$
ed+supports_pool_eio e1844) v9                                                                                                                                                                                     
2024-03-29T18:24:00.038+0000 7f4212d03700  5 --2- 127.0.0.1:0/440189136 >> [v2:127.0.0.1:6810/2525037141,v1:127.0.0.1:6811/2525037141] conn(0x1bf4630 0x1bf6a50 crc :-1 s=READY pgs=21 cs=0 l=1 rev1=1 crypto rx=0 
tx=0 comp rx=0 tx=0).write_message sending message m=0x1e6b770 seq=2940 osd_op(client.4237.0:4342 1.as0 1:5feaa37b:::o061534161-264:head [read 31956~17196,checksum in=4b] snapc 0=[] ondisk+read+known_if_redirec$ed+supports_pool_eio e1844) v9                                                                                                                                                                                     append oid (excl) 591 current snap is 14                                                                
```